### PR TITLE
Feature/cgroup resource management

### PR DESCRIPTION
**Pull request by colindrewes**
fbf02c8362526916e678be8c4451ee4dcca04eaf - Removes a path in outer.c that can never be taken. cgroupfd == -1 can never be sent from the root process to the outer process. 

0d7e47e9ab71092e5452d550265901ad21d976d1 - Removes the need to specify --cgroup which now defaults to "/sys/fs/cgroup" + what is stored in /proc/self/cgroup.

### CODE REVIEW

File: enter.c [hunk 1]
```diff
@@ -230,31 +230,55 @@ int enter(struct entry_settings *opts)
...
+
+		// If cgroup path is unspecified then attempt to derive from /proc/self/cgroup
+		if (opts->cgroup_path == NULL) {
+			FILE *selfcgroupfd = fopen("/proc/self/cgroup", "r");
+			if (selfcgroupfd == NULL) {
+				err(1, "unable to derive current cgroup hierarchy from /proc/self/cgroup");
+			}
+
+			// Move past "0:/" cgroup root in /proc/self/cgroup to get path
+			if (fseek(selfcgroupfd, 3, SEEK_SET) != 0) {
```
Comment (Snaipe) [inline enter.c:247]: This file may have multiple lines when cgroupsv1 are involved (for instance, one cgroupv1 controller is mounted while v2 is also mounted). It's better to parse this line by line and search for the "0::" prefix.

It's also possible that this is a full cgroups v1 system and there is no path. This is not a cause for error, but we should simply warn and move on as if --cgroup was not specified and was mandatory.

File: enter.c [hunk 2]
```diff
@@ -230,31 +230,55 @@ int enter(struct entry_settings *opts)
...
-		char *subcgroup = makepath("bst.%d", rootpid);
-
-		if (mkdirat(cgroupfd, subcgroup, 0777) == -1) {
-			err(1, "unable to create sub-hierachy cgroup %s", subcgroup);
-		}
-
 		// Send cleanup cgroup fd to the cleanup process
 		send_fd(outer_helper.fd, cgroupfd);
 
+		// Wait until priviliged helper creates subcgroup
```
Comment (Snaipe) [inline enter.c:274]: micro-nit: typo on privileged

File: enter.c [hunk 3]
```diff
@@ -263,8 +287,6 @@ int enter(struct entry_settings *opts)
 		apply_climits(subcgroupfd, opts->climits);
 
 		close(subcgroupfd);
-	} else if (opts->nactiveclimits != 0 && opts->cgroup_path == NULL) {
-		errx(1, "unable to apply limits without --cgroup specified");
```
Comment (Snaipe) [inline enter.c:267]: This is likely still useful when we cannot determine the root cgroup for whatever reason

File: outer.c [hunk 4]
```diff
@@ -315,6 +324,21 @@ void outer_helper_spawn(struct outer_helper *helper)
...
+		make_capable(BST_CAP_DAC_OVERRIDE);
+
+		if (mkdirat(cgroupfd, subcgroup, 0777) == -1) {
+			err(1, "unable to create sub-hierachy cgroup %s", subcgroup);
+		}
+
+		reset_capabilities();
+
+		// Notify root process cgroup hierarchy has been created
+		send_fd(fdpair[SOCKET_CHILD], cgroupfd);
```
Comment (Snaipe) [inline outer.c:340]: Rather than doing this back-and-forth, I'd argue that the privileged helper could also be the one adding the process to cgroups.procs. It's just a matter of writing `child_pid` a handful of lines below rather than `"0"` to the file.

Comment (colindrewes) [reply outer.c:340]: This choice of having the root process write to the cgroup.procs was so that if it fails to write "0" to cgroup.procs the outer helper would still cleanup the created bst.<pid> file. It seemed like this could be a pretty common occurrence on a mis-configured system. Is this not really a concern?

File: enter.c (hunk 1 above)
Comment (colindrewes) [reply enter.c]: OK. Now searches each line for prefix.

File: enter.c (hunk 2 above)
Comment (colindrewes) [reply enter.c]: Thanks. Fixed.

File: enter.c (hunk 3 above)
Comment (colindrewes) [reply enter.c:267]: Added this back in case parsing /proc/self/cgroup does not turn anything up.

File: enter.c (hunk 1 above)
Comment (colindrewes) [reply enter.c]: If there is no path to derive (on a cgroups v1 system) but there are --limits specified this should be a error right?

Comment (Snaipe) [reply enter.c]: Probably. It'll fail anyway to open the files in question since they're going to end up missing from the v2 hierarchy, so I'd be inclined to not explicitly fix it.

File: enter.c [hunk 5]
```diff
@@ -228,43 +228,54 @@ int enter(struct entry_settings *opts)
...
 		// Send cleanup cgroup fd to the cleanup process
 		send_fd(outer_helper.fd, cgroupfd);
 
-		int subcgroupfd = openat(cgroupfd, subcgroup, 0);
-		if (subcgroupfd == -1) {
-			err(1, "unable to open cgroup %s/%s", opts->cgroup_path, subcgroup);
-		}
-
-		burn(subcgroupfd, "cgroup.procs", "0");
+		// Send cleanup cgroup fd to the cleanup process
```
Comment (Snaipe) [inline enter.c:272]: comment isn't quite accurate

File: outer.c [hunk 6]
```diff
@@ -315,6 +324,34 @@ void outer_helper_spawn(struct outer_helper *helper)
 
 	if (helper->cgroup_enabled) {
 		int cgroupfd = recv_fd(fdpair[SOCKET_CHILD]);
+
+		char *subcgroup = makepath("bst.%d", rootpid);
+
+		// Writing to cgroup directory likely requires privilige
```
Comment (Snaipe) [inline outer.c:330]: micro-nit: typo

File: outer.c [hunk 7]
```diff
@@ -315,6 +324,34 @@ void outer_helper_spawn(struct outer_helper *helper)
...
+
+		reset_capabilities();
+
+		int subcgroupfd = openat(cgroupfd, subcgroup, 0);
+		if (subcgroupfd == -1) {
+			err(1, "unable to open cgroup %s", subcgroup);
+		}
+
+		char srootpid[BUFSIZ];
+		if (sprintf(srootpid, "%d", rootpid) == -1) {
```
Comment (Snaipe) [inline outer.c:345]: Is the reason why we can't move all that after child_pid is defined because we enter the cgroup namespace before?

One weakness of the current implementation is that the bst process itself counts towards the cgroup limits, rather than the child process only. It'd be nice if the child was the first process of the new cgroup.

This might require unsharing the cgroups namespace after the initial fork, so maybe not for this pull request. I'm fine if this is done separately, though.

File: enter.c [hunk 8]
```diff
@@ -544,7 +555,6 @@ int enter(struct entry_settings *opts)
 		 */
 		if (cgroupst.st_dev != 0 && cgroupst.st_dev != rootst.st_dev) {
 			const char *target = makepath("%s/sys/fs/cgroup", root);
-
```
Comment (Snaipe) [inline enter.c:547]: nit: unrelated whitespace change

File: outer.c (hunk 7 above)
Comment (colindrewes) [reply outer.c:345]: Yes that was an issue I encountered in the first draft. We need to child pid to add to the bst cgroup, but this is only accessible after the fork, which is after the ns_enter(). With an extra call to ns_enter() we can re-unshare the cgroups though.

Comment (Snaipe) [reply outer.c:345]: My proposal is to unshare the cgroup ns only once, but to do so after the fork. In other words, ns_enter shares enters all namespaces except for the cgroup namespace, which must be done post-fork and post-cgroups-setup

File: enter.c [hunk 9]
```diff
@@ -161,15 +161,55 @@ int enter(struct entry_settings *opts)
...
+			while (fgets(selfcgroup, BUFSIZ, selfcgroupfd) != NULL) {
+				if (strncmp(selfcgroup, "0::/", 3) == 0) {
+					// Remove newline character read by fgets
+					selfcgroup[strcspn(selfcgroup, "\n")] = '\0';
+					break;
+				}
+			}
+			fclose(selfcgroupfd);
+
+			// Form path from default cgroups path + current cgroup path with removed 0:/
```
Comment (Snaipe) [inline enter.c:189]: What happens if you don't find the 0::/? It looks like this just continues on with the last line of /proc/self/cgroup, and adding 3 to it, which is going to be very wrong

File: enter.c [hunk 10]
```diff
@@ -161,15 +161,55 @@ int enter(struct entry_settings *opts)
...
+		if (opts->cgroup_path == NULL) {
+			FILE *selfcgroupfd = fopen("/proc/self/cgroup", "r");
+			if (selfcgroupfd == NULL) {
+				err(1, "unable to derive current cgroup hierarchy from /proc/self/cgroup");
+			}
+
+			char selfcgroup[BUFSIZ];
+
+			while (fgets(selfcgroup, BUFSIZ, selfcgroupfd) != NULL) {
+				if (strncmp(selfcgroup, "0::/", 3) == 0) {
```
Comment (Snaipe) [inline enter.c:181]: the size argument for strncmp is supposed to be the size of the buffer, not the size of the string constant; this should be BUFSIZ, even if it seems a bit ridiculous.

The reason why is that if for some insane reason, BUFSIZ is 2, then this ends up reading past the end of the selfcgroup buffer, whereas the string constant is always well-formed, and strncmp stops as soon as its sees the null terminator of the string constant.

File: enter.c [hunk 11]
```diff
@@ -161,15 +161,55 @@ int enter(struct entry_settings *opts)
...
+					// Remove newline character read by fgets
+					selfcgroup[strcspn(selfcgroup, "\n")] = '\0';
+					break;
+				}
+			}
+			fclose(selfcgroupfd);
+
+			// Form path from default cgroups path + current cgroup path with removed 0:/
+			if (sprintf(cgroup_path, "/sys/fs/cgroup/%s", selfcgroup + 3) == -1) {
```
Comment (Snaipe) [inline enter.c:190]: Adding 3 here feels wrong; also, sprintf is dangerous since it does not check the buffer size. It might be OK right now, since both buffers are the same size, but this is fragile as if the size gets changed in one of the buffers (which is the case for my PATH_MAX comment above), it'll just introduce at best a bug, at worst a security vulnerability.

How about something like this?

```c
const char *selfcgroup = NULL;
char line[BUFSIZ];
while (fgets(line, sizeof (line), selfcgroupfd) != NULL) {
	if (strncmp(line, "0::/", sizeof (line)) == 0) {
		// Remove newline character read by fgets
		line[strcspn(line, "\n")] = '\0';
		selfcgroup = line + 3;
		break;
	}
}
fclose(selfcgroupfd);

if (selfcgroup != NULL) {
	makepath(cgroup_path, "/sys/fs/cgroup/%s", selfcgroup);
}
```

File: outer.c [hunk 12]
```diff
@@ -358,6 +346,55 @@ void outer_helper_spawn(struct outer_helper *helper)
...
+		if (sprintf(pidstr, "%d", child_pid) == -1) {
+			err(1, "outer_helper: unable to convert child_pid to string");
+		}
+
+		// Add bst subprocess to cgroup
+		burn(subcgroupfd, "cgroup.procs", pidstr);
+
+		close(subcgroupfd);
+
+		reset_capabilities();
```
Comment (Snaipe) [inline outer.c:381]: Can't this be run right after the mkdirat?

**Comment by Snaipe** [comment]
Rebase and merged at 491f60b9ca405c016146d403df5c0e04657a7152
